File: rtp_llm/server/recent_cache_key_window.py
import logging
import os
import time
from collections import Counter, deque
from dataclasses import dataclass
from typing import Callable, Deque, Dict, Iterable, Optional


DEFAULT_TIME_WINDOW_MS = 30 * 60 * 1000
CACHE_HIT_TIME_WINDOW_MS_ENV = "CACHE_HIT_TIME_WINDOW_MS"


@dataclass(frozen=True)
class RecentCacheKeySnapshot:
    time_window_ms: int
    request_occurrences: int
    request_hit_occurrences: int
    request_hit_ratio: float
    retained_occurrences: int
    retained_unique_cache_keys: int


class RecentCacheKeyWindow:
    def __init__(
        self,
        time_window_ms: Optional[int] = None,
        now_ms_supplier: Optional[Callable[[], int]] = None,
    ) -> None:
        if time_window_ms is None:
            time_window_ms = self._resolve_env_window_ms()
        self.time_window_ms = self._normalize_time_window_ms(time_window_ms)
        self._now_ms_supplier = now_ms_supplier or (lambda: int(time.time() * 1000))
        self._window_entries: Deque[tuple[int, Dict[int, int]]] = deque()
        self._cache_key_counts: Counter[int] = Counter()
        self._retained_occurrences = 0

    def record(self, cache_keys: Optional[Iterable[int]]) -> RecentCacheKeySnapshot:
        now_ms = self._now_ms_supplier()
        self._evict_expired(now_ms)

        if not cache_keys:
            return self._snapshot(0, 0)

        entry_counts: Counter[int] = Counter()
        request_occurrences = 0
        request_hit_occurrences = 0
        for cache_key in cache_keys:
            if cache_key is None:
                continue
            request_occurrences += 1
            if cache_key in self._cache_key_counts:
                request_hit_occurrences += 1
            entry_counts[int(cache_key)] += 1

        if not entry_counts:
            return self._snapshot(0, 0)

        self._window_entries.append((now_ms, dict(entry_counts)))
        self._cache_key_counts.update(entry_counts)
        self._retained_occurrences += sum(entry_counts.values())
        return self._snapshot(request_occurrences, request_hit_occurrences)

    def snapshot(self) -> RecentCacheKeySnapshot:
        self._evict_expired(self._now_ms_supplier())
        return self._snapshot(0, 0)

    def _evict_expired(self, now_ms: int) -> None:
        expire_before_or_at = now_ms - self.time_window_ms
        while self._window_entries:
            timestamp_ms, entry_counts = self._window_entries[0]
            if timestamp_ms > expire_before_or_at:
                return
            self._window_entries.popleft()
            for cache_key, expired_count in entry_counts.items():
                current_count = self._cache_key_counts.get(cache_key, 0)
                if current_count <= expired_count:
                    self._cache_key_counts.pop(cache_key, None)
                    self._retained_occurrences -= current_count
                else:
                    self._cache_key_counts[cache_key] = current_count - expired_count
                    self._retained_occurrences -= expired_count

    def _snapshot(
        self, request_occurrences: int, request_hit_occurrences: int
    ) -> RecentCacheKeySnapshot:
        hit_ratio = (
            request_hit_occurrences / request_occurrences
            if request_occurrences > 0
            else 0.0
        )
        return RecentCacheKeySnapshot(
            time_window_ms=self.time_window_ms,
            request_occurrences=request_occurrences,
            request_hit_occurrences=request_hit_occurrences,
            request_hit_ratio=hit_ratio,
            retained_occurrences=self._retained_occurrences,
            retained_unique_cache_keys=len(self._cache_key_counts),
        )

    @staticmethod
    def _resolve_env_window_ms() -> int:
        value = os.environ.get(CACHE_HIT_TIME_WINDOW_MS_ENV)
        if value is None:
            return DEFAULT_TIME_WINDOW_MS
        try:
            return int(value)
        except ValueError:
            logging.warning(
                "Invalid %s=%s, fallback to default: %s",
                CACHE_HIT_TIME_WINDOW_MS_ENV,
                value,
                DEFAULT_TIME_WINDOW_MS,
            )
            return DEFAULT_TIME_WINDOW_MS

    @staticmethod
    def _normalize_time_window_ms(candidate_ms: int) -> int:
        if candidate_ms > 0:
            return candidate_ms
        logging.warning(
            "Invalid cache hit time window: %s, fallback to default: %s",
            candidate_ms,
            DEFAULT_TIME_WINDOW_MS,
        )
        return DEFAULT_TIME_WINDOW_MS

File: rtp_llm/cpp/cache/RecentCacheKeyWindow.cc
#include "rtp_llm/cpp/cache/RecentCacheKeyWindow.h"

#include <algorithm>
#include <cstdlib>
#include <exception>
#include <string>
#include <utility>

#include "rtp_llm/cpp/utils/Logger.h"
#include "rtp_llm/cpp/utils/TimeUtil.h"

namespace rtp_llm {
namespace {

constexpr const char* kPrefillCacheHitTimeWindowMsEnv = "PREFILL_CACHE_HIT_TIME_WINDOW_MS";
constexpr const char* kCacheHitTimeWindowMsEnv        = "CACHE_HIT_TIME_WINDOW_MS";

int64_t normalizeTimeWindowMs(int64_t candidate_ms) {
    if (candidate_ms > 0) {
        return candidate_ms;
    }
    RTP_LLM_LOG_WARNING("Invalid cache hit time window ms: %ld, fallback to default: %ld",
                        candidate_ms,
                        RecentCacheKeyWindow::DEFAULT_TIME_WINDOW_MS);
    return RecentCacheKeyWindow::DEFAULT_TIME_WINDOW_MS;
}

bool parseEnvMs(const char* env_name, int64_t& value) {
    const char* raw = std::getenv(env_name);
    if (raw == nullptr || raw[0] == '\0') {
        return false;
    }
    try {
        value = std::stoll(std::string(raw));
        return true;
    } catch (const std::exception& e) {
        RTP_LLM_LOG_WARNING("Invalid %s=%s, fallback to default: %ld, error=%s",
                            env_name,
                            raw,
                            RecentCacheKeyWindow::DEFAULT_TIME_WINDOW_MS,
                            e.what());
        value = RecentCacheKeyWindow::DEFAULT_TIME_WINDOW_MS;
        return true;
    }
}

}  // namespace

RecentCacheKeyWindow::RecentCacheKeyWindow():
    RecentCacheKeyWindow(resolveTimeWindowMsFromEnv(), []() { return currentTimeMs(); }) {}

RecentCacheKeyWindow::RecentCacheKeyWindow(int64_t time_window_ms, NowSupplier now_supplier):
    time_window_ms_(normalizeTimeWindowMs(time_window_ms)), now_supplier_(std::move(now_supplier)) {
    if (!now_supplier_) {
        now_supplier_ = []() { return currentTimeMs(); };
    }
}

RecentCacheKeyWindow::Snapshot RecentCacheKeyWindow::record(const std::vector<CacheKeyType>& cache_keys) {
    std::lock_guard<std::mutex> lock(mutex_);
    const int64_t               now_ms = now_supplier_();
    evictExpiredLocked(now_ms);

    if (cache_keys.empty()) {
        return snapshotLocked(0, 0);
    }

    std::unordered_map<CacheKeyType, int64_t> entry_counts;
    entry_counts.reserve(cache_keys.size());

    int64_t request_occurrences     = 0;
    int64_t request_hit_occurrences = 0;
    for (const auto cache_key : cache_keys) {
        ++request_occurrences;
        if (cache_key_counts_.find(cache_key) != cache_key_counts_.end()) {
            ++request_hit_occurrences;
        }
        ++entry_counts[cache_key];
    }

    if (entry_counts.empty()) {
        return snapshotLocked(0, 0);
    }

    WindowEntry entry;
    entry.timestamp_ms     = now_ms;
    entry.cache_key_counts = std::move(entry_counts);

    for (const auto& [cache_key, count] : entry.cache_key_counts) {
        cache_key_counts_[cache_key] += count;
        retained_occurrences_ += count;
    }
    window_entries_.push_back(std::move(entry));

    return snapshotLocked(request_occurrences, request_hit_occurrences);
}

RecentCacheKeyWindow::Snapshot RecentCacheKeyWindow::snapshot() {
    std::lock_guard<std::mutex> lock(mutex_);
    evictExpiredLocked(now_supplier_());
    return snapshotLocked(0, 0);
}

int64_t RecentCacheKeyWindow::timeWindowMs() const {
    std::lock_guard<std::mutex> lock(mutex_);
    return time_window_ms_;
}

int64_t RecentCacheKeyWindow::resolveTimeWindowMsFromEnv() {
    int64_t value = 0;
    if (parseEnvMs(kPrefillCacheHitTimeWindowMsEnv, value) || parseEnvMs(kCacheHitTimeWindowMsEnv, value)) {
        return normalizeTimeWindowMs(value);
    }
    return DEFAULT_TIME_WINDOW_MS;
}

void RecentCacheKeyWindow::evictExpiredLocked(int64_t now_ms) {
    const int64_t expire_before_or_at = now_ms - time_window_ms_;
    while (!window_entries_.empty() && window_entries_.front().timestamp_ms <= expire_before_or_at) {
        auto entry = std::move(window_entries_.front());
        window_entries_.pop_front();
        for (const auto& [cache_key, expired_count] : entry.cache_key_counts) {
            auto it = cache_key_counts_.find(cache_key);
            if (it == cache_key_counts_.end()) {
                continue;
            }
            const int64_t decrement = std::min(it->second, expired_count);
            retained_occurrences_ -= decrement;
            if (it->second <= expired_count) {
                cache_key_counts_.erase(it);
            } else {
                it->second -= expired_count;
            }
        }
    }
}

RecentCacheKeyWindow::Snapshot RecentCacheKeyWindow::snapshotLocked(int64_t request_occurrences,
                                                                    int64_t request_hit_occurrences) const {
    Snapshot snapshot;
    snapshot.time_window_ms          = time_window_ms_;
    snapshot.request_occurrences     = request_occurrences;
    snapshot.request_hit_occurrences = request_hit_occurrences;
    snapshot.request_hit_ratio =
        request_occurrences > 0 ? static_cast<double>(request_hit_occurrences) / request_occurrences : 0.0;
    snapshot.retained_occurrences       = retained_occurrences_;
    snapshot.retained_unique_cache_keys = cache_key_counts_.size();
    return snapshot;
}

}  // namespace rtp_llm
